Datasources: Improve error handling for error messages

ifrost opened the pull request:
**What this PR does / why we need it**:

This PR improves the way we handle errors when testing a data source results in failure (when clicking on "Test & Save" button).

So far the logic was to return `error.statusText` and fall back to `error.message` if no `statusText` is defined. In practice `error.statusText` is a more generic message and `error.message` should provide more detailed information. What is more the error might be returned in the response body in `error.data.message`. This PR changes that the error message is extracted from the first non-empty property in the following order:

- `error.message`
- `error.data.message`
- `error.statusText`

This should give more precise information about what happened.

<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #34013

Backend changed the logic so in case of authentication errors  HTTP 401 is no longer returned but more appropriate in this case HTTP 400 with a message in the response body (see the issue for the explanation). This, however, means that the logic described above is picking up the status code for 400 (Bad Request) without additional info. By changing the way the error message is extracted we will get a more meaningful error.

**How to test it?**

Run `make devenv sources=graphite`. Create a Graphite data source. For given URLs you should get the following errors:

| URL | Error message |
|-----|----------------|
| https://graphite-us-central1.grafana.net/graphite | Authentication to data source failed |
| http://bad-gateway | Bad Gateway | 
| http://localhost:8180 | Data source is working |

for example:

<img width="660" alt="Screenshot 2021-06-02 at 17 21 54" src="https://user-images.githubusercontent.com/745532/120507423-16998480-c3c7-11eb-911c-b24dc6d49b2f.png">


**Special notes for your reviewer**:

N/A

ifrost committed:
Improve error handling for error messages

The error message will be read from error object from the following properties in the following order:
- message
- data.message
- statusText

marefr reviewed (changes requested):
Works great for Graphite, but not for other data sources. Would have been nice to include this in v8, but seems like too less time for that?

For example just trying to connect to Microsoft SQL Server data source without having it running I get this message:
![image](https://user-images.githubusercontent.com/1668778/120657597-f03a1e80-c484-11eb-8d98-da3801e9ce66.png)

Response payload looks like this:
```json
{
  "results": {
    "A": {
      "error": "db query error: failed to connect to server - please inspect Grafana server log for details",
      "frames": [
        {
          "schema": {
            "meta": { "executedQueryString": "SELECT 1" },
            "fields": []
          },
          "data": { "values": [] }
        }
      ]
    }
  }
}

```

And the MSSQL datasource then converting that response here:
https://github.com/grafana/grafana/blob/1c49986b2f2e36179159258517696353a5ea01c8/public/app/plugins/datasource/mssql/datasource.ts#L170-L177

This probably applies to any SQL data source and would guess basically any backend data source. For example all backend data sources basically using this code:
https://github.com/grafana/grafana/blob/1c49986b2f2e36179159258517696353a5ea01c8/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L209-L217

ifrost commented:
Thanks @marefr! For Microsoft SQL Server I'm getting "Metric request error" (which is also not super helpful). MySQL is indeed returning just "400" though it's not related to changes in this PR but returned but the datasource API ([testDatasource()](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/mysql/datasource.ts#L191)). I just changed how exceptions thrown by `testDatasource` are handled.

It seems like the only way to fix all data sources is to check every data source one by one (probably also enterprise plugins). Might be also useful to update docs on how errors should be handled by data sources.

This PR still should improve the API in general and at least fix it for Graphite. Do you think it's worth merging it and then work on all data sources separately?

hugohaggmark reviewed (commented):
I agree with @marefr and maybe we should use or create a similar function to `toDataQueryError` if that can't handle this scenario.
https://github.com/grafana/grafana/blob/43d3d9756285caec2621707471f7e08a86f6771e/packages/grafana-runtime/src/utils/queryResponse.ts#L136

ifrost committed:
Merge branch 'main' into ifrost/improve-testing-ds-error-handling

ifrost committed:
Merge branch 'main' into ifrost/improve-testing-ds-error-handling

ifrost committed:
Convert api/ds/query errors to TestingStatus

SQL datasources (mysql, mssql, postgres) and CloudWatch use api/ds/query to test the data source, but previously didn't handle errors returned by this endpoint. If the error cannot be handled it's re-thrown to be handled in public/app/features/datasources/state/actions.ts

ifrost commented:
With 86cfd40 I added error handling for SQL datasources (mysql, mssql, postgres) and CloudWatch - all use `api/ds/query` to test the data source. Other errors (including HTTP errors) are re-thrown to be handled in `public/app/features/datasources/state/actions.ts`.

`toTestingStatus()` handles 2 types of errors returned by `api/ds/query`. Both can be tested with Microsoft Server SQL:
- create a new data source and click save (`api/ds/query` will return an object with `{ message, error }`)
- add database name and click save (`api/ds/query` will return empty query results with an error in `results.A.error`)

Out of curiosity - in ideal world, should all DataSourcesWithBackend implement health check endpoint and use it instead?

hugohaggmark commented on public/app/plugins/datasource/cloudwatch/datasource.ts:
```diff
@@ -881,10 +882,14 @@ export class CloudWatchDatasource extends DataSourceWithBackend<CloudWatchQuery,
-      status: 'success',
-      message: 'Data source is working',
-    }));
+    return this.getDimensionValues(region, namespace, metricName, 'ServiceName', dimensions)
+      .then(() => ({
+        status: 'success',
+        message: 'Data source is working',
+      }))
+      .catch((error) => {
+        return toTestingStatus(error);
+      });
```
nit: can we use async/await instead?

ifrost replied:
Good point! Updated with 6b5efd6

hugohaggmark reviewed (approved):
Such splendid work on this, love that you added tests and comments. I haven't got the time to test it locally yet so I'll approve this for frontend platform. Maybe the individual data sources need to approve this works as expected.

ifrost committed:
Merge branch 'main' into ifrost/improve-testing-ds-error-handling

ifrost committed:
Use async/await instead of Promises

ifrost committed:
Merge branch 'main' into ifrost/improve-testing-ds-error-handling

ifrost committed:
Merge branch 'main' into ifrost/improve-testing-ds-error-handling

ifrost committed:
Remove incorrect type import

TestingStatus is in app/types. Should be pulled down to grafana/data but it depends on HealthCheckResultDetails that is public and lives in grafana/runtime. Ideally TestingStatus should live in grafana/data but I'm not sure if HealthCheckResultDetails can be move there too (?)

ifrost commented:
I had to remove returned value for `TestingStatus` in 1bd4266. `TestingStatus` lives in [`app/types/datasources`](https://github.com/grafana/grafana/blob/main/public/app/types/datasources.ts#L19) so cannot be imported directly to packages. Should it be moved somewhere else? `@grafana/data` or `@grafana/runtime` packages? On one hand it's returned by [`@grafana/data: DataSourceApi`](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/datasource.ts#L233) but on the other hand requires [`@grafana/runtime: HealthCheckResultDetails`](https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L62).

Any guidelines on where types for data sources should be defined?

ifrost commented:
> Works great for Graphite, but not for other data sources. Would have been nice to include this in v8, but seems like too less time for that?

@marefr I've changed the implementation to cover SQL data sources, I hope it's better now 👍

ifrost commented:
@sunker could you have a quick look if changes for CloudWatch make sense?

marefr commented on packages/grafana-data/src/types/datasource.ts:
```diff
@@ -472,7 +476,13 @@ export enum DataQueryErrorType {
+     */
     message?: string;
+    /**
+     * Detailed information about the error
```
```suggestion
     * Detailed information about the error. Only returned when app_mode is development.
```

ifrost replied:
@marefr 👍 Added. Though I was getting it also when running in production mode, so maybe *sometimes* it's returned?

marefr replied:
Hmm are you sure? I had this code in mind https://github.com/grafana/grafana/blob/4632b070e4a4c736698df61c18494902a9f7d209/pkg/api/response/response.go#L194. Possible that this is not used for all responses I suppose?

ifrost replied:
Lemme check again 🔎

ifrost replied:
Oh sorry, my bad! You're totally right - `data.error` is not returned in production mode 🤦 As discussed [below](https://github.com/grafana/grafana/pull/35120#discussion_r663755749) I have changed the logic to cover this case. Thanks!

marefr commented on packages/grafana-runtime/src/utils/queryResponse.ts:
```diff
@@ -127,6 +127,36 @@ export function toDataQueryResponse(
+export function toTestingStatus(err: FetchError): any {
+  const queryResponse = toDataQueryResponse(err);
+  // POST api/ds/query errors returned as { message: string, error: string } objects
+  if (queryResponse.error?.data?.message && queryResponse.error?.data?.error) {
```
I think this logic is wrong since error is only included when [app_mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode) is development. But I might be wrong since not sure if `toDataQueryResponse` somehow handles this case and always populates the `data.error`?

```suggestion
  if (queryResponse.error?.data?.message) {
```

marefr commented on packages/grafana-runtime/src/utils/queryResponse.ts:
```diff
@@ -127,6 +127,36 @@ export function toDataQueryResponse(
+    return {
+      status: 'error',
+      message: queryResponse.error.data.message,
+      details: { message: queryResponse.error.data.error },
```
Based on earlier suggestion the details might not be included if `queryResponse.error.data.error` is undefined?

ifrost replied:
@marefr Good catch. Though, I think the end result would be the same because when `data.error` is not defined then `toTestingStatus` re-throws the error and the message is extracted from the error in `public/app/features/datasources/state/actions.ts`:

```typescript
const message = errMessage || data?.message || 'HTTP error ' + statusText;
```

I have changed it so when `data.error` is missing it's handled earlier, in `toTestingStatus`, it's probably more readable this way.

marefr reviewed (approved):
I think this looks good :+1: I've tested to connect to almost all core data sources without having their backend services running and seems to work as expected. Still a bit unsure if we're missing something for external plugins, but given they're returning `status` and `message` already it should be fine. Only possible case would be an unhandled exception, but guess it should fallback to `statusText`. I'll request a review from Ryan/Torkel just in case I've missed something.

Added some suggestions.

sunker commented on packages/grafana-data/src/types/datasource.ts:
```diff
@@ -228,7 +228,11 @@ abstract class DataSourceApi<
-   * Test & verify datasource settings & connection details
+   * Test & verify datasource settings & connection details (returning TestingStatus)
+   *
+   * When verification fails - errors specific to the data source should handled here and converted to
```
```suggestion
   * When verification fails - errors specific to the data source should be handled here and converted to
```

sunker reviewed (approved):
Nice! Tested CloudWatch, Cloud Monitoring and Azure Monitor and it works as expected.

ifrost committed:
Merge branch 'main' into ifrost/improve-testing-ds-error-handling

ifrost committed:
Update packages/grafana-data/src/types/datasource.ts

Co-authored-by: Erik Sundell <erik.sundell@grafana.com>

ifrost committed:
Handle errors with no details in toTestingStatus instead of re-throwing

ifrost committed:
Merge branch 'ifrost/improve-testing-ds-error-handling' of https://github.com/grafana/grafana into ifrost/improve-testing-ds-error-handling

ifrost committed:
Update packages/grafana-data/src/types/datasource.ts

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

ifrost commented:
> (...). I'll request a review from Ryan/Torkel just in case I've missed something.

@torkelo  / @ryantxu - could you have a quick look if these changes make sense?

ryantxu reviewed (approved):
LGTM

ifrost committed:
Merge branch 'main' into ifrost/improve-testing-ds-error-handling

# Conflicts:
#	public/app/plugins/datasource/cloudwatch/datasource.ts

Dieterbe commented:
thanks @ifrost :heart_eyes:
